parser.go

```diff
@@ -0,0 +1,345 @@
+	err := configStandardCallbacks(&ret, fieldVal)
+	if err == nil {
+		return paramName, ret, nil
+	}
+
+	// try to configure it as an "xtype"
+	if isXType(structField.Type) {
+		if fieldVal.IsNil() {
+			fieldVal.Set(reflect.New(fieldVal.Type().Elem()))
+		}
+
+		ret.boolean = describeType(fieldVal) == "bool"
+
+		ret.validFn = toXType(fieldVal).ValueValid
+		ret.setValueFn = toXType(fieldVal).UnmarshalParam
+		ret.getDefaultFn = toXType(fieldVal).GetDefaultValue
+
+		// some types know how to redact themselves (for example,
+		// xtype.URL know how to redact the password)
+		if redactor := toRedactor(fieldVal); redactor != nil {
+			ret.redactFn = redactor.RedactValue
+		} else {
+			ret.redactFn = func(s string) string { return s }
+		}
+
+		return paramName, ret, nil
+	}
+
+	// if is a struct, assume it to be a parameter set
+	if fieldVal.Kind() == reflect.Struct {
+		ret.paramSet = true
+
+		// parameter sets have no value, and the callback functions should
+		// not be called; install handlers to help debug in case of a mistake.
+		panicMessage := fmt.Sprintf("%q is a paramset, it have no value", paramName)
+		ret.validFn = func(v string) error { panic(panicMessage) }
+		ret.setValueFn = func(v *string) error { panic(panicMessage) }
+		ret.getDefaultFn = func() (string, error) { panic(panicMessage) }
+		ret.redactFn = func(s string) string { panic(panicMessage) }
+
+		return paramName, ret, nil
+	}
+
+	return paramName, ret, fmt.Errorf("struct member %q is unsupported", paramName)
+}
+
+func addSpecialFlags(appConfig config, parsed *Parsed, opts settings) error {
+	// support special flags
+	var violations types.ErrViolations
+	if opts.autoUsageExitFn != nil {
+		if conflictingParam, ok := appConfig[""].fields[specialflags.Help.Name]; ok {
+			violations = append(violations, types.Violation{
+				ParamName: specialflags.Help.Name,
+				Path:      conflictingParam.path,
+				Message:   "The help parameter cannot be used with the auto-usage is requested",
+			})
+		}
+
+		appConfig[""].fields[specialflags.Help.Name] = paramSetField{
+			typ:      "bool",
+			optional: true,
+			desc:     specialflags.Help.Description,
+			boolean:  true,
+
+			setValueFn: func(v *string) error {
+				parsed.Usage(os.Stdout)
+				opts.autoUsageExitFn()
+				return nil
+			},
+
+			validFn:      func(v string) error { return nil },
+			getDefaultFn: func() (string, error) { return "false", nil },
+			redactFn:     func(s string) string { return s },
+		}
+	}
+
+	if len(violations) > 0 {
+		return nil
+	}
+
+	return nil
+}
+
+func processSpecialFlags(appConfig config, parsed *Parsed, opts settings) {
+	if parsed.settings.autoUsageExitFn == nil || parsed.readValue("", specialflags.Help.Name) == nil {
+		return
+	}
+
+	fmt.Fprintln(opts.autoUsageWriter, opts.autoUsageHeadline)
+	parsed.Usage(opts.autoUsageWriter)
+	parsed.settings.autoUsageExitFn()
+	panic("Auto usage termination callback function did not terminated the application")
```

Comment:
I really like to have callers of the library being written like:

```go
params := struct{...}{...}

parsed, err := proteus.MustParse(&params, 
		proteus.WithAutoUsage(os.Stdout, "My Test App", func() { os.Exit(0) }),
		...)
if err != nil {
	parsed.ErrUsage(os.Stderr, err)
	os.Exit(1)
}
// more code
```

For dealing with your concern regarding panicking when the termination function does not terminate, I tried a few alternatives:

- returning an error from `MustParse`: this is bad because it will result in usage information being shown; the output ends up being confusing
- printing a warning, then terminating with status 0: this works quite nice, even better then in the case of panic, the output is more readable and the code is more appropriate

So I changed to the second option (terminating with status 0).

Testing the `--help` flag can still be accomplished by providing an exit function that panics with a value that is known by the test, and then recovering on the test code. It is not beautiful, but it follows the principle of _simple things are easy, complicated things are possible_ (simple being the code snippet above, complicated being the test case for the help flag).